src/js/client/editprefset-viewport.js

```diff
@@ -0,0 +1,681 @@
+ * set as `null`.
+ */
+gpii.devpmt.lookupGenericPrefValue = function (that, context, commonTerm) {
+    if (that.model.flatPrefs.contexts[context].preferences[commonTerm] !== undefined) {
+        return that.model.flatPrefs.contexts[context].preferences[commonTerm];
+    }
+    else {
+        return undefined;
+    }
+};
+
+/**
+ * Lookup Application Specific Setting. Similar to `lookupGenericPrefValue` lookups
+ * in regard to return values, but takes an extra argument for the product.
+ */
+gpii.devpmt.lookupProductPrefValue = function (that, context, product, settingTerm) {
+    if (that.model.flatPrefs.contexts[context].preferences[product] &&
+        that.model.flatPrefs.contexts[context].preferences[product][settingTerm] !== undefined) {
+        return that.model.flatPrefs.contexts[context].preferences[product][settingTerm];
+    }
+    else {
+        return undefined;
+    }
+};
+
+gpii.devpmt.initSettingTableWidgets = function (that) {
+    var existingTables = fluid.queryIoCSelector(fluid.rootComponent, "gpii.devpmt.settingsTableWidget");
+    fluid.each(existingTables, function (item) {
+        item.destroy();
+    });
+    fluid.each(that.model.npsetApplications, function (item) {
+        var sel = "#test-comp-" + item.appId;
+        sel = sel.replace(/\./g, "\\.");
+        var appUri = "http://registry.gpii.net/applications/" + item.appId;
+        that.events.createSettingsTable.fire(sel, item.appId, that.model.allSolutions[item.appId], appUri);
+    });
+};
+
+gpii.devpmt.addEditToUnsavedList = function (that, description) { // that, path, newValue, oldValue) {
+    that.model.unsavedChanges.push({
+        // path: path,
+        // newValue: newValue,
+        // oldValue: oldValue
+        description: description
+    });
+};
+
+gpii.devpmt.updateFoundationSticky = function () {
+    // https://github.com/zurb/foundation-sites/issues/7899
+    $(".sticky").foundation("_calc", true);
+};
+
+/**
+ * When the prefs are updated we need to usually update
+ * some of the metadata that makes rendering easier, from
+ * the new preference information.
+ *
+ * This should maybe be a set of model relay rules instead.
+ */
+gpii.devpmt.updateMetadataFromPrefs = function (that) {
+    that.applier.change("contextNames", gpii.devpmt.contextNames(that.model.flatPrefs));
+};
+
+gpii.devpmt.openAddProductDialog = function (that, appId) {
+    that.applier.change("activeModalDialog", {
+        appId: appId,
+        name: that.model.allSolutions[appId].name
+    });
+    that.events.openConfirmAddProductDialog.fire();
+};
+
+gpii.devpmt.savePrefset = function (that /*, event */) {
+    var options = {
+        method: "POST",
+        contentType: "application/json",
+        url: "/saveprefset/" + that.model.npsetName,
+        data: JSON.stringify({ flat: that.model.flatPrefs }, null, 4)
+    };
+    $.ajax(options);
+    that.applier.change("unsavedChangesExist", false);
```

Comment:
You might want to wrap this pair of updates in a transaction - or else apply them at the root

Reply:
These are wrapped in a transaction now.